### ISSUE TITLE
Extensions cannot set DYLD_LIBARARY_PATH on spawned processes.

### ISSUE THREAD

**Issue by joe-jordan** [microsoft/vscode#88306]
Apologies, I can update these when I get back to the office.
- VSCode Version: Latest stable channel
- OS Version: macOS

Steps to Reproduce:

1. Install the gtest adapter extension from the store.
2. Load a C++ project which sets DYLD_LIBRARY_PATH so that its gtest unit test executable can find the debug builds of its dynamic libraries.
3. Run the launch target from vs code itself - the exe runs fine.
4. Try to refresh the list of tests on the gtest adapter GUI - it fails.

I am trying to use the gtest extension for vs code. This extension needs to execute my tests binary as the debugger would, but with a different argument (e.g. `--gtest-list-tests` or similar) in order to show me a list of tests to choose one to run/debug.

When the extension tries to execute this (using `child_process.execSync`) it is not permitted by macOS to set the DYLD_LIBRARY_PATH variable, as specified in my launch configuration, which means it cannot run the process correctly.

Your main debugger is allowed to set this variable, and is able to launch my unit tests (although obviously it runs all of them at once, so debugging is much less convenient).

This is a macOS security setting that the main debugger must have the exemption for, but the extension process seems not to. Please give the permission to the extension process!

You can read about the security setting here:

https://developer.apple.com/documentation/security/hardened_runtime_entitlements

The relevant section is headed "Allow DYLD Environment Variables Entitlement".

See also - the bug I raised with Node.js (which they resolved as not valid, since it's a macOS "feature".):

https://github.com/nodejs/node/issues/31184

Does this issue occur when all extensions are disabled?: Yes/No Not Applicable - it relates to the operation of extensions.

**Comment by joe-jordan** [comment]
I apologise - I wasn't clear enough in my first report. Let me try again:

VSCode about dialog:
- Version: 1.41.1
- Commit: 26076a4de974ead31f97692a0d32f90d735645c0
- Date: 2019-12-18T14:57:51.166Z
- Electron: 6.1.5
- Chrome: 76.0.3809.146
- Node.js: 12.4.0
- V8: 7.6.303.31-electron.0
- OS: Darwin x64 17.5.0

This issue is not caused by an extension, it is an issue with the subsystem which executes extension code.

Here is the extension for which it is obvious:

https://github.com/DavidSchuldenfrei/gtest-adapter

My pull request to the extension to fix an unrelated bug:

https://github.com/DavidSchuldenfrei/gtest-adapter/pull/48

I have spend some time debugging this extension on both Linux and macOS. My debugging has involved building a dummy binary which logs arguments and environment variables to see whether the relevant env vars are visible to it. My conclusions are:
- On Linux, the extension behaves correctly.
- On macOS, the extension behaves incorrectly.
- The extension contains no platform-specific code, except that which checks for platform-specific settings in the launch configurations (which is simply about ensuring the correct settings are applied on each platform.)

Please see my original report for details of the incorrect behaviour.

I have tracked this bug down with the help of the Node.js team (it is reproducible in the Homebrew Node.js, which has the same problem as the visual studio code extension runner - please see the linked issue on the node.js repo in my original report above).

I have found that it is because of a build time setting of visual studio code, specifically the binary which becomes the process which executes extensions. The component of VS Code which executes debug processes *is* permitted to pass through DYLD environment variables to its subprocesses - however the extension runner component is not (perhaps this is a deliberate security decision on your part - I would be interested to know.)

The details of the build setting can be found here:

https://developer.apple.com/documentation/security/hardened_runtime_entitlements

under the section  "Allow DYLD Environment Variables Entitlement".

I humbly request that you reopen the issue and begin an investigation or discussion about whether your security policy permits giving extensions the ability to pass through DYLD environment variables to their subprocesses. 

(Or, perhaps if the API for this already exists, point me to some documentation which shows me the correct way to execute subprocesses from an extension so that such variables can be passed through, so that I can develop the relevant patch in my fork of the extension and submit it as a pull request to the extension maintainer.)

Thank you,

Joe Jordan

**Comment by weinand** [comment]
I'm not aware of any special treatment for DYLD_LIBRARY_PATH permissions in VS Code.
But be aware that extensions are run in Electron's builtin version of node.js.

What do you mean by "main debugger" in your statement "Your main debugger is allowed to set this variable"?

/cc @bpasero @deepak1556 any advice from your side?

**Comment by joe-jordan** [comment]
Main debugger: when you hit "debug", on the main debug tab, it will
correctly set the environment variables specified in the "environment" key
in the launch configuration, and they will make it to the spawned process.

When an extension does the same thing through this electron node process,
it won't be allowed to.

It's possible you don't set this permission elsewhere, and macOS is
inconsistent about where it applies the security feature (which would be
even more of a headache), but I would argue probably still a vscode bug,
since you could theoretically control it with build settings.

Thank you for reopening.

**Comment by AustinShalit** [comment]
I opened an issue, https://github.com/microsoft/vscode-java-debug/issues/742, about a week ago which I think is the same issue. I encourage folks to try the simple test cases there but to summarize, on macOS system integrity protection makes it so the DYLD_LIBRARY_PATH & LD_LIBRARY_PATH environment variables are not propagated to subprocesses. The work around we discovered is to run the debugger twice. The first time it will fail to load the environment variables. The second time, the environment variables are passed directly and the debugging works as intended.

**Comment by AustinShalit** [comment]
@weinand Do you think what I am describing is the same issue or should I open a different issue?

**Comment by deepak1556** [comment]
@joe-jordan @AustinShalit thanks for the reports, we just enabled notarization in the insiders build since last week, this particular entitlement for allowing DYLD env variables is unfortunately not set, we will enable it and check if it fixes the issue.

/cc @joaomoreno @bpasero

**Comment by deepak1556** [comment]
Can you try the latest insiders to see if this issue is fixed ? Thanks!

**Comment by AustinShalit** [comment]
The latest insiders edition works for me! Thank you for fixing this!

**Comment by weinand** [comment]
Thanks for verifying.
